tools/tpm2_pcrallocate.c

```diff
@@ -0,0 +1,241 @@
+                break;
+            case TPM2_ALG_SHA256:
+                tpm2_tool_output ("  - bank: sha256\n");
+                break;
+            case TPM2_ALG_SHA384:
+                tpm2_tool_output ("  - bank: sha384\n");
+                break;
+            case TPM2_ALG_SHA512:
+                tpm2_tool_output ("  - bank: sha512\n");
+                break;
+            default:
+                tpm2_tool_output ("  - bank: alg0x%08x\n",
+                                  pcrselection[i].hash);
+        }
+        first = true;
+        for (j = 0; j < pcrselection[i].sizeofSelect * 8; j++) {
+            if (pcrselection[i].pcrSelect[j / 8] & 1<<(j % 8)) {
+                if (first) {
+                    tpm2_tool_output ("    pcrs: [ %i", j);
+                    first = false;
+                } else {
+                    tpm2_tool_output(", %i", j);
+                }
+            }
+        }
+        tpm2_tool_output (" ]\n");
+    }
+}
+
+static bool pcr_allocate(ESYS_CONTEXT *ectx) {
+    TSS2_RC rval;
+    TPMI_YES_NO allocationSuccess;
+    UINT32 maxPCR;
+    UINT32 sizeNeeded;
+    UINT32 sizeAvailable;
+
+    rval = Esys_TR_SetAuth(ectx, ESYS_TR_RH_PLATFORM, &ctx.platauth);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_ERR("Could not set the password.");
+        LOG_PERR(Esys_TR_SetAuth, rval);
+        return false;
+    }
+
+    dump_pcrselection(ctx.pcrSelection.pcrSelections, ctx.pcrSelection.count);
+
+    rval = Esys_PCR_Allocate(ectx, ESYS_TR_RH_PLATFORM, 
+                             ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
+                             &ctx.pcrSelection, &allocationSuccess,
+                             &maxPCR, &sizeNeeded, &sizeAvailable);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_ERR("Could not allocate PCRs.");
+        LOG_PERR(Esys_PCR_Allocate, rval);
+        return false;
+    }
+
+    if (!allocationSuccess) {
+        LOG_ERR("Allocation failed. Size Needed: %i, Size available: %i",
+                sizeNeeded, sizeAvailable);
+        return false;
+    }
+    (void)(maxPCR);
+
+    return true;
+}
+
+static bool on_arg(int argc, char** argv){
+    int i;
+    uint32_t pcr;
+
+    for (i = 0; i < argc; i++){
+        if (tpm2_util_string_to_uint32(argv[i], &pcr)) {
+            if (ctx.pcrSelection.count == 0) {
+                LOG_ERR("Got PCR Index without Bank selection");
+                return false;
+            }
+
+            if (pcr >= TPM2_MAX_PCRS) {
+                LOG_ERR("Got out of bound PCR Index: \"%s\"", argv[i]);
+                return false;
+            }
+
+            if (pcr > 24)
+                ctx.pcrSelection.pcrSelections[ctx.pcrSelection.count - 1].
+                    sizeofSelect = 4;
+
+            ctx.pcrSelection.pcrSelections[ctx.pcrSelection.count - 1].
+                pcrSelect[pcr / 8] |= 1 << (pcr % 8);
+
+            continue;
+        }
+
+        if (argv[i][strlen(argv[i])-1] == ':') {
```

Comment:
Why not using `pcr_parse_selections` from `lib/pcr.h` and extend it to support the "all" keyword ?

Reply:
ACK

Reply:
Done